### ISSUE TITLE
CacheTransactionHandler's Committed() should use interceptionContext

### ISSUE THREAD

**Issue by dandzina** [moozzyk/EFCache#52]
When a SqlTransaction is committed, it's Connection property is set to null.
In the Committed method of CacheTransactionHandler, this implementation passes a null for connection to ResolveCache:

ResolveCache(transaction.Connection).InvalidateSets(entitySets.Distinct());

It should be replaced with
ResolveCache(interceptionContext.Connection).InvalidateSets(entitySets.Distinct());

which will derived classes of CacheTransactionHandler to rely on the Connection property when providing a cache.

**Comment by moozzyk** [comment]
Thanks for reporting. Indeed this seems to be the case. The documentation of `SqlTransaction.Connection` implies this by saying: "Gets the SqlConnection object associated with the transaction, or null if the transaction is no longer valid." 
Have you found it by trying to write custom cache resolution?

**Comment by dandzina** [comment]
Yes I found this when creating a derived class of CacheTransactionHandler where I attempted to just override GetCache().  I have a system where we have many databases with the same schema, and wanted to supply a separate cache to each based on connection.Database.

 

I have worked around this for now by directly implementing IDBTransactinInterceptor() and using ‘new’ modifier overrides for Committed and Rollback and otherwise overrideing every virtual method that of the class.  My override of Committed has been working well with the use of connection from the interceptionContext.  

public new void Committed(DbTransaction transaction, DbTransactionInterceptionContext interceptionContext)

    {

      var entitySets = RemoveAffectedEntitySets(transaction);

 

      if (entitySets != null)

      {

        ResolveCache(interceptionContext.Connection).InvalidateSets(entitySets.Distinct());

      }

    }

 

If your implementation is revised, I can slim down my derived class back to just GetCache as original intended.  

 

Thanks for the consideration.
